flang/lib/Optimizer/Transforms/CUFAddConstructor.cpp

```diff
@@ -106,8 +106,7 @@ struct CUFAddConstructor
 
         mlir::func::FuncOp func;
         switch (attr.getValue()) {
-        case cuf::DataAttribute::Device:
-        case cuf::DataAttribute::Constant: {
+        case cuf::DataAttribute::Device: {
```

Comment:
What about variable like this:

```
module mod1
  integer, constant :: a(10)
end
```

Do they need to be registered?

Reply:
Something like this:
```
module devmod
    use iso_c_binding, only: c_int
    integer, parameter :: amazon(10) = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    integer, constant :: google(10)
```
gets lowered to, in HLFIR:
```
  fir.global @_QMdevmodECamazon(dense<0> : tensor<10xi32>) constant : !fir.array<10xi32>
  fir.global @_QMdevmodEgoogle {data_attr = #cuf.cuda<constant>} : !fir.array<10xi32> {
```

I am not finding a whole lot about the `constant` attribute. Is it new? From online research, it seems to just be  a protection against modification such as a pointer to constant data. If it is the case, then it should be implemented like a regular global. 

The array parameter seems to be handled fine with this change. It will get the `#cuf.cuda<constant>` assigned later in the flow and will be copied to the GPU module with its initializer.

Reply:
It is defined in the programming guide: https://docs.nvidia.com/hpc-sdk/compilers/cuda-fortran-prog-guide/index.html#cfpg-var-qual-attr-constant

Maybe we wrongly add the ` #cuf.cuda<constant>` in the implicit global pass.

Reply:
```
Device constant data may not be assigned or modified in any device subprogram, but may be modified in host subprogram
```
It seems to be more like a global with some constraints on the device side.

Reply:
Yeah but shouldn't we register them as well?

Reply:
Yes. So, we should either revisit how `#cuf.cuda<constant>` is applied or how constant fir.global are handled. Whichever change we are going to apply is going to ripple through the flow. What would be the best way to handle it? Maybe we only apply the attribute to reflect the user setting?

Reply:
> Maybe we only apply the attribute to reflect the user setting?

Yeah it's probably nicer and if we need we can add an `implicit` attribute or just default to `device`.  I guess `#cuf.cuda<constant>` and `constant` in fir.global are not really synonym. We can handle this is a follow up patch when we have a real test with the CUDA attribute constant.